File: app/src/androidTest/java/me/varlez/movies/common/rest/MockMovieService.java
package me.varlez.movies.common.rest;

import java.util.ArrayList;
import java.util.List;

import me.varlez.movies.common.model.Movie;
import me.varlez.movies.common.model.SearchResults;
import retrofit2.Call;
import retrofit2.http.Query;
import retrofit2.mock.Calls;

/**
 * Mock of our movie service REST API.
 */
public class MockMovieService implements MovieService {

    private List<Movie> movies;

    public MockMovieService() {
        movies = new ArrayList<>();
        createMovies();
    }

    @Override
    public Call<Movie> movie(@Query("i") String movieId) {
        return Calls.response(movies.get(0));
    }

    @Override
    public Call<SearchResults> list(@Query("s") String query, @Query("type") String type) {
        return Calls.response(new SearchResults(movies));
    }

    private void createMovies() {
        Movie firstMovie = new Movie();
        firstMovie.setId("1");
        firstMovie.setTitle("Batman");
        firstMovie.setYear(2005);
        firstMovie.setPlot("The Batman !");
        firstMovie.setDirector("Christopher Nolan");

        Movie secondMovie = new Movie();
        secondMovie.setId("2");
        secondMovie.setTitle("Space Jam");
        secondMovie.setYear(1996);
        secondMovie.setPlot("The best movie !");
        secondMovie.setDirector("Joe Pytka");

        movies.add(firstMovie);
        movies.add(secondMovie);
    }
}

File: app/src/androidTest/java/me/varlez/movies/common/rest/MockMovieService.kt
package me.varlez.movies.common.rest

import io.reactivex.Observable
import me.varlez.movies.common.model.Movie
import me.varlez.movies.common.model.SearchResults
import retrofit2.http.Query

/**
 * Mock of our movie service REST API.
 */
class MockMovieService : MovieService {

    private val movies: MutableList<Movie> = createMovies()

    override fun movie(@Query("i") movieId: String): Observable<Movie> {
        return Observable.just(movies[0])
    }

    override fun list(@Query("s") query: String?, @Query("type") type: String): Observable<SearchResults> {
        return Observable.just(SearchResults(movies))
    }

    private fun createMovies(): ArrayList<Movie> {
        val firstMovie = Movie(id = "1",
                title = "Batman",
                year = 2005,
                posterUrl = "",
                plot = "The Batman !",
                director = "Christopher Nolan")

        val secondMovie = Movie(id = "2",
                title = "Space Jam",
                year = 1996,
                posterUrl = "",
                plot = "The best movie !",
                director = "Joe Pytka")

        return arrayListOf(firstMovie, secondMovie)
    }
}
